cms/djangoapps/contentstore/courseware_index.py

```diff
@@ -0,0 +1,196 @@
+                for child_item in item.get_children():
+                    index_item(child_item, current_start_date, skip_index=skip_child_index)
+
+            if skip_index or not item_index_dictionary:
+                return
+
+            item_index = {}
+            # if it has something to add to the index, then add it
+            try:
+                item_index.update(location_info)
+                item_index.update(item_index_dictionary)
+                item_index['id'] = item_id
+                if current_start_date:
+                    item_index['start_date'] = current_start_date
+
+                searcher.index(DOCUMENT_TYPE, item_index)
+                indexed_count["count"] += 1
+            except Exception as err:  # pylint: disable=broad-except
+                # broad exception so that index operation does not fail on one item of many
+                log.warning('Could not index item: %s - %r', item.location, err)
+                error_list.append(_('Could not index item: {}').format(item.location))
+
+        def remove_deleted_items():
+            """
+            remove any item that is present in the search index that is not present in updated list of indexed items
+            as we find items we can shorten the set of items to keep
+            """
+            response = searcher.search(
+                doc_type=DOCUMENT_TYPE,
+                field_dictionary={"course": unicode(course_key)},
+                exclude_ids=indexed_items
+            )
+            result_ids = [result["data"]["id"] for result in response["results"]]
+            for result_id in result_ids:
+                searcher.remove(DOCUMENT_TYPE, result_id)
+
+        try:
+            with modulestore.branch_setting(ModuleStoreEnum.RevisionOption.published_only):
+                course = modulestore.get_course(course_key, depth=None)
+                for item in course.get_children():
+                    index_item(item, course.start)
+                remove_deleted_items()
+        except Exception as err:  # pylint: disable=broad-except
+            # broad exception so that index operation does not prevent the rest of the application from working
+            log.exception(
+                "Indexing error encountered, courseware index may be out of date %s - %r",
+                course_key,
+                err
+            )
+            error_list.append(_('General indexing error occurred'))
+
+        if error_list:
+            raise SearchIndexingError(_('Error(s) present during indexing'), error_list)
+
+        return indexed_count["count"]
+
+    @classmethod
+    def do_course_reindex(cls, modulestore, course_key):
+        """
+        (Re)index all content within the given course, tracking the fact that a full reindex has taking place
+        """
+        indexer = cls()
+        indexed_count = indexer.index_course(modulestore, course_key)
```

Comment:
@ormsbee there's a [related PR](https://github.com/edx/edx-platform/pull/7448) that adds content libraries indexing. It was submitted earlier than this one, but there is an irremovable defect with tests: search indexers used to live in `common/lib`, which should not depend on any django features - but search indexers implicitly depends on `SEARCH_ENGINE` setting. So, since this PR significantly changes and improves indexing *and* fixes that test issue, it was decided #7448 is suspended until this PR is merged.

So, having indexers a classmethods or instance methods was my suggestion to simplify rebasing #7448 on top of this. @martynjames likely chosen instance methods because I used instance methods in #7448 (but he definitely knows better). If it makes any difference, I used instance methods in #7448 because I personally find classes with instance methods induce less coupling between "provider" and "client" code. Static and class methods explicitly mentions implementing class, while instance methods can be called on any class implementing certain inteface, thus allowing swapping implementations easier than with static/class methods.